### ISSUE TITLE
map editor

### ISSUE THREAD

**Issue by Scutterman** [Scutterman/ascii-lemmings-wasm#60]
The solution to our problem of not being able to store all of our map data in a human readable way is the same solution games developers have been using for decades - have a map editor to convert between ugly format and viewable format.

**Comment by Scutterman** [comment]
We want the level editor to have different controls:
- toggle layer button (toggle between editing the map and background)
- Click on each tile to edit it
- Must scoll like the full level
- Text box to enter tile
- Text box to enter animation frames
- Text box to enter colour

**Comment by Scutterman** [comment]
Notes from tonight:

Each level should have the following metadata:
- Animation List
- Default animations

Each tile may have the following custom animation metadata:
- Animation list key

To edit a tile:
- Click on tile
- Add border to tile
- Show tile animation list key
- Allow changing tile animation list key

On tile animation change:
- Update meta map
- Update map

Each tile is made up from a tile in the map and possibly a custom animation
When changing a tile, we should be able to change the tile type and optionally give it a new custom animation

Once a tile is selected, two buttons can be clicked.
- "Update Tile" allows selecting of a tile type
- "Update Animation" allows selecting of an animation list item key

We have most of the above done, except clicking the Tile and Animation buttons doesn't seem to do anything right now.

**Comment by Scutterman** [comment]
Map scrolling doesn't trigger map re-render, we currently have to wait until a tile is modified.

**Comment by Scutterman** [comment]
Display each character of each animation frame as a button.
Clicking on the button allows typing a different character.

**Comment by Scutterman** [comment]
We have a "new row" and a "new column" button in place, but they need finishing off.

When adding a row or column to the map editor, the blocks for that row / column need to be added into the html as divs and the css as .row_x and .col_y

**Comment by Scutterman** [comment]
Need to add a "New level" button beside the "load" button

**Comment by Scutterman** [comment]
Need to save the map into our custom format and use assemblyscript code generation to turn it into a class at compile time.

**Comment by Scutterman** [comment]
Map won't load into map editor
Map editor needs to generate `//EDITORHINT::MAP_METADATA` followed by a map name line when it's generating the map file
Map editor needs the ability to set the skills up and then add that to the generated map file

Animation editor needs to generate the global animation file when the save button is clicked.
If all characters of an animation are the same and there's only one frame, it should be a single character animation, otherwise a full animation.

**Comment by Scutterman** [comment]
Animation editor needs to generate the global animation file when the save button is clicked.
If all characters of an animation are the same and there's only one frame, it should be a single character animation, otherwise a full animation.

**Comment by Scutterman** [comment]
Map editor needs to be able to set the default animations for each map

**Comment by Scutterman** [comment]
Map editor needs to be able to set number to spawn and number for success

**Comment by Scutterman** [comment]
Map editor needs the ability to set the skills up and then add that to the generated map file

**Comment by Scutterman** [comment]
Map editor needs the ability to set level difficulty, number, and code when creating a new level

**Comment by Scutterman** [comment]
Need to add spawn, success, and skill counts to LevelMetadata class and parsers

**Comment by Scutterman** [comment]
When adding a new map, the "add row", "add col" and "save" buttons do not function

**Comment by Scutterman** [comment]
When restarting the level, the skills don't get repopulated. This is probably due to something to do with the way we're parsing the level and the way we're cloning it. Probably best to look at this after we work out how we're adding skills to meta or map.

**Comment by Scutterman** [comment]
There's no way to share code between typescript and assemblyscript due to missing features and differing return types. Specifically, `Array.from()` is missing and `Map.keys()` returns a `string[]` not an iterator.

As a result, we've had to remove parsing from the Edit screen. Our options to solve this are either:

1. Load levels in using similar code to the title screen and give up the functionality where users could load in user-generated content
2. See if we've managed to get good enough separation between the `scaffolding` folder and the `assemblyscript` folder that we can move the `scaffolding` back to using assemblyscript transforms.

I like the second option, but don't spend too much time on it.
